### PR TITLE
issue #1636:  start CheckRunner if check exists in consul already

### PR DESCRIPTION
**Pull request by capone212**
CheckRunner currently starts only during check registration in registerCheck() method. This is a problem when check already exists in consul (nomad executor registered it during previews run). In that case registerCheck is not called for a check and nobody calls .Start() method of CheckRunner. In result we have check in TTL expired state

### CODE REVIEW

**Comment by dadgar** [comment]
Ah nice catch! Did this fix your issue?

**Comment by capone212** [comment]
yes. One that relates to failing checks. I still reproduce the over, with duplicating allocation for job. I am planing to attack it tomorrow.

File: command/agent/consul/syncer.go [hunk 1]
```diff
@@ -684,6 +687,20 @@ func (c *Syncer) registerCheck(chkReg *consul.AgentCheckRegistration) error {
...
 
+// ensureCheckRunning verifies that registerd check is in running state
+func (c *Syncer) ensureCheckRunning(chk *consul.AgentCheckRegistration) {
+	c.registryLock.RLock()
+	defer c.registryLock.RUnlock()
+	cr, ok := c.checkRunners[consulCheckID(chk.ID)]
+	if !ok {
+		return
+	}
+	if !cr.Started() {
```
Comment (diptanu) [inline command/agent/consul/syncer.go]: `if ok && !cr.Started()` and remove the `!ok` check from above.

File: command/agent/consul/syncer.go [hunk 2]
```diff
@@ -684,6 +687,20 @@ func (c *Syncer) registerCheck(chkReg *consul.AgentCheckRegistration) error {
 	return c.client.Agent().CheckRegister(chkReg)
 }
 
+// ensureCheckRunning verifies that registerd check is in running state
+func (c *Syncer) ensureCheckRunning(chk *consul.AgentCheckRegistration) {
+	c.registryLock.RLock()
+	defer c.registryLock.RUnlock()
+	cr, ok := c.checkRunners[consulCheckID(chk.ID)]
```
Comment (diptanu) [inline command/agent/consul/syncer.go]: I think we can simplify this -

```if cr, ok := c.checkRunners[consulCheckID(chk.ID)]; ok && !cr.Started() {
 		c.logger.Printf("[DEBUG] consul.syncer: starting runner for existing check. %v", chk.ID)
 		cr.Start()
 	}
```

Note that I have also suggested changing the log message slightly.

File: command/agent/consul/check.go [hunk 3]
```diff
@@ -47,6 +47,12 @@ func (r *CheckRunner) Start() {
 	r.started = true
 }
 
+func (r *CheckRunner) Started() bool {
```
Comment (diptanu) [inline command/agent/consul/check.go]: Please add a comment.

**Comment by diptanu** [comment]
@capone212 Thanks! Added a few comments.

File: command/agent/consul/check.go (hunk 3 above)
Comment (capone212) [inline command/agent/consul/check.go]: Do you think the following will be ok?
// Started is used to check whether Start() method called for this object
excuse for my English

Comment (diptanu) [inline command/agent/consul/check.go]: The comment could be -
```
Started returns if the check runner has started running
```

File: command/agent/consul/syncer.go [hunk 4]
```diff
@@ -684,6 +687,20 @@ func (c *Syncer) registerCheck(chkReg *consul.AgentCheckRegistration) error {
 	return c.client.Agent().CheckRegister(chkReg)
 }
 
+// ensureCheckRunning verifies that registerd check is in running state
```
Comment (diptanu) [inline command/agent/consul/syncer.go]:
```
// ensureCheckRunning starts the check runner for a check if it's not already running
```

**Comment by capone212** [comment]
@diptanu thanks for suggestions! updated PR

**Pull request merged**